hardsigmoid: add cuda kernels

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36420 hardswish: add backards pass test
* **#36351 hardsigmoid: add cuda kernels**

Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20955589](https://our.internmc.facebook.com/intern/diff/D20955589)

vkuzo committed:
hardsigmoid: add cuda kernels

Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "hardsigmoid: add cuda kernels"


Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20955589](https://our.internmc.facebook.com/intern/diff/D20955589)

[ghstack-poisoned]

vkuzo committed:
Update on "hardsigmoid: add cuda kernels"


Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20955589](https://our.internmc.facebook.com/intern/diff/D20955589)

[ghstack-poisoned]

vkuzo committed:
Update on "hardsigmoid: add cuda kernels"


Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20955589](https://our.internmc.facebook.com/intern/diff/D20955589)

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -522,6 +522,38 @@ void hardswish_backward_kernel(TensorIterator& iter) {
 }
 
+void hardsigmoid_kernel(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "hardsigmoid_cuda", [&]() {
```
Tensorflow uses a different definition, https://www.tensorflow.org/api_docs/python/tf/keras/activations/hard_sigmoid,
we do not have a use case for that definition, so lets stick with this one

vkuzo committed:
Update on "hardsigmoid: add cuda kernels"


Summary:

Adds CUDA kernels for hardsigmoid, to enable its use in training.

Note: the update to the cpu backward pass is to keep the cpu vs cuda
logic consistent, no change in functionality.

Test Plan:

add CI for the forward pass
run this for the backward pass:
https://gist.github.com/vkuzo/95957d365600f9ad10d25bd20f58cc1a

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20955589](https://our.internmc.facebook.com/intern/diff/D20955589)

[ghstack-poisoned]
